symengine/numerics/math.h

```diff
@@ -0,0 +1,732 @@
+{
+    // Call other implementation
+    return max(GenericExpression<EnableImplicitConstructor>(a), b);
+}
+
+/**
+ * Return a symbolic number that represents the minimum of two
+ * values @p a and @p b.
+ *
+ * Mimics the function <code> std::min(a,b) </code> using the standard math
+ * library.
+ */
+template<bool EnableImplicitConstructor>
+GenericExpression<EnableImplicitConstructor> min(const GenericExpression<EnableImplicitConstructor> &a, const GenericExpression<EnableImplicitConstructor> &b);
+
+/**
+ * Return a symbolic number that represents the minimum of two
+ * values @p a and @p b.
+ *
+ * Mimics the function <code> std::min(a,b) </code> using the standard math
+ * library.
+ *
+ * This variant is used when @p b is not a GenericExpression<EnableImplicitConstructor>.
+ */
+template <
+bool EnableImplicitConstructor,
+    typename NumberType,
+    typename = typename std::
+        enable_if<!std::is_same<NumberType, GenericExpression<EnableImplicitConstructor>>::value>::type>
+GenericExpression<EnableImplicitConstructor> min(const GenericExpression<EnableImplicitConstructor> &a, const NumberType &b)
+{
+    // Call other implementation
+    return min(a, GenericExpression<EnableImplicitConstructor>(b));
+}
+
+/**
+ * Return a symbolic number that represents the minimum of two
+ * values @p a and @p b.
+ *
+ * Mimics the function <code> std::min(a,b) </code> using the standard math
+ * library.
+ *
+ * This variant is used when @p a is not a GenericExpression<EnableImplicitConstructor>.
+ */
+template <
+bool EnableImplicitConstructor,
+    typename NumberType,
+    typename = typename std::
+        enable_if<!std::is_same<NumberType, GenericExpression<EnableImplicitConstructor>>::value>::type>
+GenericExpression<EnableImplicitConstructor> min(const NumberType &a, const GenericExpression<EnableImplicitConstructor> &b)
+{
+    // Call other implementation
+    return min(GenericExpression<EnableImplicitConstructor>(a), b);
+}
+
+/**
+ * Return a symbolic number that represents error function with the
+ * given argument @p x.
+ *
+ * Mimics the function <code> std::erf(x) </code> using the standard math
+ * library.
+ */
+template<bool EnableImplicitConstructor>
+GenericExpression<EnableImplicitConstructor> erf(const GenericExpression<EnableImplicitConstructor> &x);
+
+/**
+ * Return a symbolic number that represents complimentary error function
+ * with the given argument @p x.
+ *
+ * Mimics the function <code> std::erfc(x) </code> using the standard math
+ * library.
+ */
+template<bool EnableImplicitConstructor>
+GenericExpression<EnableImplicitConstructor> erfc(const GenericExpression<EnableImplicitConstructor> &x);
+
+//@}
+
+} // namespace Numerics
+} // namespace SymEngine
+
+#ifndef DOXYGEN
+
+// Import math operations into standard namespace.
+// This gives one the ability to use them within more libraries with generic
+// templated number types.
+namespace std
```

Comment:
I'm not a fan of this. I think the `std` namespace should be left as is. Why is this necessary?

Reply:
For example, Eigen has the following about a scalar type `T`

> The math function should be defined in the same namespace than T, or in the std namespace though that second approach is not recommended.

Reply:
There are other libraries that export their math functions to the `std` namespace. For example, [Trilinos Sacado](https://github.com/trilinos/Trilinos/blob/master/packages/sacado/src/new_design/Sacado_Fad_Exp_MathFunctions.hpp) does this, and its very convenient for reusing existing templated code (originally written with arithmetic types in mind) with these SymEngine numbers without further adaption.

We could do this import locally (as we have done [here](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/adolc_math.h) for ADOL-C numbers), but this is (a) annoying, (b) would likely be repeated by other libraries and therefore (c) runs the risk of having duplicate function definitions between cross-linked libraries, which would be difficult to resolve. 

I strongly believe that its a convenient and useful thing to do. To make this optional, one could guard this with a compile-time activated `#define`.

Reply:
> We could do this import locally (as we have done [here](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/adolc_math.h) for ADOL-C numbers), but this is (a) annoying, (b) would likely be repeated by other libraries and therefore (c) runs the risk of having duplicate function definitions between cross-linked libraries, which would be difficult to resolve.

I agree that doing this in dealii or other libraries is messy. I just don't understand where these are needed though. Do you have an example where these functions are used in dealii?

Reply:
Here's [one example](https://github.com/dealii/dealii/blob/master/include/deal.II/base/symmetric_tensor.h#L2250-L2285) in out tensor library. There must be countless other examples in the code base, but that's the first obvious one that I could find. Most of our users would write boundary conditions and constitutive models with this basic syntax
```
template<typename T = double>
T func (T x){
  return std::sin(x)*std::cos(x);
}
```
and
```
template<typename T = double, int dim>
T neohookean_energy_function (SymmetricTensor<2/*rank*/,dim,T> C){
  return 0.5*(trace(C) - dim - std::log(determinant(C));
}
```
rather than
```
template<typename T = double>
T func (T x){
  using std::sin;
   using std::cos;
  return sin(x)*cos(x);
}
```
and
```
template<typename T = double, int dim>
T neohookean_energy_function (SymmetricTensor<2/*rank*/,dim,T> C){
  using std::log;
  return 0.5*(trace(C) - dim - log(determinant(C));
}
```

I would speculate that this is the same across most numerical codes that primarily think in terms of arithmetic types (both libraries and user codes). So importing these math functions into the `std` namespace would really be a huge help to those that want to use `SymEngine` as a "drop-in" replacement of sorts that is compatible with existing code bases.

Reply:
Okay, thanks for the example. Now I understand why an implicit constructor with int, float types would be bad, but I still don't like having two types (one with the implicit constructors and the one without)

How about declaring,
```
Expression std::sin(RCP<const Basic> &x)
```
instead of 
```
Expression std::sin(Expression &x)
```

This should fix the problem of conflicting with int and float types, right?

Reply:
I don't think that this would work in general. As far as I could tell, this wouldn't allow compound operations like `std::sin(std::sqrt(x))`. Or am I missing something?

Reply:
Note that adding new overloads to ``std`` is not only bad practice, it is explicitly forbidden by the standard and technically it results in undefined behaviour (and yes, I am aware that it tends to work in practice).

Reply:
Fair enough. If that's the case then I guess we'd consider another approach: We would then provide our users with a stand-alone header that does this and that they can optionally include themselves.

Reply:
To be fair, non-operator customisation points are a sore point in C++, and the fact that people keep on feeling the impulse to put stuff in ``std`` is one of the symptoms of that :)

What I usually end up doing is to approach the problem in the opposite way: instead of trying to add new overloads/specialisations to ``std::`` functions (say, ``sin()``), I define a ``sin()`` function in one of my namespaces, provide implementations of ``sin(float)``/``sin(double)`` based on the ``std::`` functions and then allow the users to overload/specialise my implementation for their types via  ADL, partial template specialisation, overloading, etc., depending on the situation. There's no silver bullet though (or at least, I have not found it), and different solutions come with different tradeoffs in terms of complexity, efficiency, etc.

http://ericniebler.com/2014/10/21/customization-point-design-in-c11-and-beyond/
https://quuxplusone.github.io/blog/2018/03/19/customization-points-for-functions/

Reply:
> it is explicitly forbidden by the standard and technically it results in undefined behaviour

Just for the record, this is exactly what is explained at https://en.cppreference.com/w/cpp/language/extending_std . Thanks for the info!

> then allow the users to overload/specialise my implementation for their types via ADL

Thanks as well for this tip. A couple of years ago I discussed the use of ADL versus overloads in `std` when starting to extend our library to support arbitrary number types. The problem that we ran into is that we have ~1 million line codebase, and doing this systematically as required was simply too burdensome. But I do agree that relying on ADL would be the most correct way to do this. Having all of the math operators in `SymEngine::Numerics` would of course permit this. 

In summary, I'd be OK with removing these imports from SymEngine and putting it into our own codebase.

Reply:
> I don't think that this would work in general. As far as I could tell, this wouldn't permit compound operations like `std::sin(std::sqrt(x))`. Or am I missing something?

It would because `Expression` is implicitly convertible to `RCP<const Basic>`

Reply:
With the current implementation this is not the case:
```
    // Conversion to SymEngine types
    {
      const Expression x("x");
      const SE::RCP<SE::Basic> r1 (x);
      const SE::RCP<SE::Basic> r2 = x;
    }
```
leads to 
```
<path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:148:32: error: no viable conversion from 'const Expression' (aka 'const GenericExpression<false>') to 'const SE::RCP<SE::Basic>'
      const SE::RCP<SE::Basic> r2 = x;
                               ^    ~
<path>/symengine-dev/symengine/utilities/teuchos/Teuchos_RCPDecl.hpp:459:10: note: candidate constructor not viable: no known conversion from 'const Expression' (aka 'const GenericExpression<false>') to 'Teuchos::ENull' for 1st argument
  inline RCP(ENull null_arg = null);
         ^
<path>/symengine-dev/symengine/utilities/teuchos/Teuchos_RCPDecl.hpp:524:10: note: candidate constructor not viable: no known conversion from 'const Expression' (aka 'const GenericExpression<false>') to 'const RCP<SymEngine::Basic> &' for 1st argument
  inline RCP(const RCP<T>& r_ptr);
         ^
<path>/symengine-dev/symengine/utilities/teuchos/Teuchos_RCPDecl.hpp:540:10: note: candidate template ignored: could not match 'RCP' against 'GenericExpression'
  inline RCP(const RCP<T2>& r_ptr);
         ^
In file included from <path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:3:
<path>/symengine-dev/symengine/numerics/number_types.h:1313:22: error: implicit instantiation of undefined template 'SymEngine::Numerics::internal::Evaluator<Teuchos::RCP<SymEngine::Basic>, void>'
    return internal::Evaluator<ResultType>::value(get_RCP());
                     ^
<path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:147:36: note: in instantiation of function template specialization 'SymEngine::Numerics::GenericExpression<false>::operator RCP<Teuchos::RCP<SymEngine::Basic> >' requested here
      const SE::RCP<SE::Basic> r1 (x);
                                   ^
<path>/symengine-dev/symengine/numerics/utilities.h:288:8: note: template is declared here
struct Evaluator;
       ^
In file included from <path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:3:
<path>/symengine-dev/symengine/numerics/number_types.h:1313:43: error: incomplete definition of type 'SymEngine::Numerics::internal::Evaluator<Teuchos::RCP<SymEngine::Basic>, void>'
    return internal::Evaluator<ResultType>::value(get_RCP());
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
3 errors generated.
```
It might be possible to implement conversion through `operator=`, but I don't see how its possible to do this for the constructor `SE::RCP<SE::Basic> (Expression)`

Reply:
> It would because `Expression` is implicitly convertible to `RCP<const Basic>`

Could you perhaps elaborate on how this was done in the current `SymEngine::Expression` class, so that I could replicate it?

Reply:
You are right. I only implemented it in https://github.com/symengine/symengine/pull/1507. See https://github.com/symengine/symengine/pull/1507/files#diff-5052f081551b80163e1096fe0a85fcefR134

Reply:
Cool, thanks. I think that this might work. I need to remove all of the templates to `GenericExpression`, and will let you know the outcome once I've done that.